### ISSUE TITLE
Timestamp issues on MacBook Air M1

### ISSUE THREAD

**Issue by SRSaunders** [KhronosGroup/MoltenVK#1438]
After a bit of investigation, I have found two things that may be causing problems with timestamps on Apple Silicon:

1. On my MacBook Air M1, using SDK 1.2.189.0, `vkcontext.gpu->props.limits.timestampPeriod = 41.6667`.  Looking at Sascha Willems' Vulkan hardware database (https://vulkan.gpuinfo.org), earlier SDKs showed a `timestampPeriod = 1` for M1 machines.  Since the RBDOOM-3-BFG game uses `timestampPeriod` to calibrate its timing intervals, this would lead to longer than expected times in the stats counter for the game.  This is what I am seeing.  Has something changed regarding `timestampPeriod` in the newer MoltenVK builds?

2. A second issue on the M1 appears to be revealed by calling `vkCmdWriteTimestamp()` back-to-back with no additional commands in between.  There is at least one situation in the RBDOOM-3-BFG game when a rendering stage is skipped for Vulkan (but not for OpenGL), and the following commands are issued:
```
  vkCmdWriteTimestamp( commandBuffer, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, queryPool, queryIndex1 );

  <...some trivial cpu logic, but no additional GPU commands when using the Vulkan renderer...>

  vkCmdWriteTimestamp( commandBuffer, VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT, queryPool, queryIndex2 );

  <...and then later on...>

  vkGetQueryPoolResults( vkcontext.device, queryPool, MRB_GPU_TIME, numQueries, results.ByteSize(), results.Ptr(), sizeof( uint64 ), VK_QUERY_RESULT_64_BIT | VK_QUERY_RESULT_WAIT_BIT );
```
For my Intel Mac with Radeon 580 graphics, MoltenVK correctly returns identical timestamps when queried later (i.e. zero time difference).  However, on the M1 this appears to be a problem and the returned timestamps are different, leading to a incorrect non-zero time interval for this situation.

**Comment by billhollings** [comment]
Thanks for submitting this.

1. The `41.6667` value is accurate on `arm64` for CPU-based time-stamping (without GPU counters). But that is useful only for older iOS environments. With GPU-counters, it is indeed inaccurate and needs to be fixed. The challenge when using GPU counters is that there is no fixed `timestampPeriod`, as the GPU's can run at different rates, depending on load, and so the raw timestamp period will fluctuate over time. We'll have to figure out how to normalize that to a fixed `timestampPeriod`.

2. Interesting. On M1, the timestamps should occur when Metal encoding is finished, typically at the end of a render subpass, and so we should expect that two consecutive timestamps with nothing in-between should be identical. What's happening around the code above? Are both those `vkCmdWriteTimestamp()` commands part of a single renderpass? We just need to understand the conditions those calls are being made so we can look into it.

**Comment by SRSaunders** [comment]
Thanks for looking into the `timestampPeriod` issue.  With variable GPU rates it sounds like a tricky problem.  I hope you can come up with a solution for a nominal value.

Regarding the consecutive timestamp issue, I have been digging deeper into the problem by collecting data across multiple runs through a single renderpass (containing only 1 subpass with multiple rendering steps queued up by CPU) on both my Intel machine+Radeon 580 and my Apple Silicon M1.  Comparing the two data sets shows some interesting info:

1. On the Radeon 580 the run counts are extremely consistent, both in terms of raw timestamp intervals within timestamp pairs (where work is done in a rendering step) and gaps between timestamp pairs (between rendering steps).  Within a pair, intervals are non-zero when there is work, and zero when there is no work (e.g. if timestamps are requested but no rendering commands placed in between).  In addition the gaps are zero between consecutive timestamp pairs within a single renderpass, i.e. in-between rendering steps when the GPU should not be doing anything.

2. On the M1, the raw run counts are less consistent, possibly due to the variable rate GPU that you mention above.  However, it is interesting to note that not only are the intervals within a pair non-zero when there is no work, but also the gaps between rendering steps are non-zero, when the GPU is not supposed to be doing work.  It's almost as if the M1's GPU timestamps are using a global time counter vs. a command-relative time counter in the Radeon 580, or possibly if current timestamps on the M1 are not fully synced to the work they surround in the command buffer.

I am not sure if this information helps.  If you would like the raw timestamp data for comparison, I could link or email an excel sheet to you.  Please let me know.

**Comment by billhollings** [comment]
>If you would like the raw timestamp data for comparison, I could link or email an excel sheet to you. Please let me know.

If it's not too much trouble to provide the spreadsheet, that might be helpful. If you would prefer to directly email it, rather than post it publicly here, please email it to support@brenwill.com.

**Comment by SRSaunders** [comment]
Excel sheet emailed.  Please let me know if you have any questions.  I would be happy to provide additional info if I can.

**Comment by billhollings** [comment]
@SRSaunders 

I’ve been digging into the various issues here. 

Do you have frame-over-frame numbers for M1? I want to confirm that on M1 the GPU counter numbers are already in nanoseconds, in which case `timestampPeriod` should be changed to `1` on M1. Frame-over-frame GPU counter values at a known frame rate would confirm that.

Or if you have already confirmed that in your testing, just let me know here.

**Comment by SRSaunders** [comment]
I just measured the frame-over-frame numbers on the M1 using the MRB_GPU_TIME start marker (ref the spreadsheet I sent) and found that the calibration is indeed in nanoseconds (i.e. timestampPeriod = 1).  I was able to set the game frame rate to a fixed 60 FPS and also to a fixed 120 FPS for comparison.  The measured frame-over-frame period using the M1 GPU counters with a timestampPeriod of 1 yields approx 16 msec and 8 msec respectively, as would be expected for those frame rates.

However, there is still something wrong with the measured intra-frame times for the various rendering steps as described in my spreadsheet.  Even with a corrected timestampPeriod = 1, the various Start marker to End marker rendering periods within a frame appear too small.

**Comment by billhollings** [comment]
PR #1500 contains a partial fix for this.

That PR updates the value of `VkPhysicalDeviceLimits:: timestampPeriod` to be 1.0 for all Apple GPUs when using GPU Counters, and for non-Apple GPUs, it dynamically adjusts the value for `VkPhysicalDeviceLimits:: timestampPeriod` by correlating the GPU and CPU timestamps.

Unfortunately, there doesn't seem to be anything obvious to be done regarding GPU time passing when the GPU is not doing anything. This behavior is a side effect of the fact that on Apple Silicon, the GPU and CPU use the same realtime nanosecond timestamps, so GPU timestamps will advance even if no GPU activity is taking place. These GPU times will typically be very small if no GPU activity occurs, but will generally not be zero.

**Comment by SRSaunders** [comment]
@billhollings 

Thanks for working on this.  I looked at the changes and wonder if there is something missing in the update calculation.  In `MVKPhysicalDevice::updateTimestampPeriod`, you update the GPU timestamp period based on the CPU to GPU counter difference ratio, but don't take into consideration the actual period of the CPU ticks.  Are you assuming that CPU ticks are always = 1 nanosecond (ok for Intel CPUs), but is this correct for other situations?  I notice that you skip the update calculation for Apple GPUs, and thus infer and skip for Apple CPUs as well.  The current calculation looks like this:

_properties.limits.timestampPeriod = (double)(cpuEnd - cpuStart) / (double)(gpuEnd - gpuStart);

In the general case, shouldn't the code read something like this instead:

_properties.limits.timestampPeriod = ( (double)(cpuEnd - cpuStart) / (double)(gpuEnd - gpuStart) ) * mvkGetTimestampPeriod();

Independent of the above, I am pretty confused about why `mvkGetTimestampPeriod()` returns 41.67 (125/3) for Apple Silicon, where CPU ticks appear to be in nanoseconds on the M1 chip as evidenced by your code and empirically from frame-over-frame measurements I have done previously on this thread.  Is there something I don't understand or off with the return value of `mach_timebase_info()` on M1 machines?

**Comment by billhollings** [comment]
> Are you assuming that CPU ticks are always = 1 nanosecond (ok for Intel CPUs), but is this correct for other situations?

Apple's [documentation](https://developer.apple.com/documentation/metal/counter_sampling/correlating_cpu_and_gpu_timestamps?language=objc) indicates that for `sampleTimestamps:gpuTimestamp:`

> The system measures CPU timestamps in nanoseconds.

So that function seems to be internally taking into consideration the clock tick interval, and the results appear in nanoseconds already.

---

> why `mvkGetTimestampPeriod()` returns 41.67 (125/3) for Apple Silicon

All Apple CPUs seem to operate on this clock tick interval. 

On the M1 system I have here, if I set the following environment variables:

```
MVK_CONFIG_PERFORMANCE_TRACKING=1
MVK_CONFIG_PERFORMANCE_LOGGING_FRAME_COUNT=60
```

and then run the Cube demo, it logs a steady 60.0 FPS and 15ms frame times. These calculations use `mvkGetTimestampPeriod()` and the logged results would be wildly different if off by a factor of 41.67.

In summary, the M1 clock tick indeed seems to be 41.67 ns, and Apple accommodates for this and publishes the CPU results of `sampleTimestamps:gpuTimestamp:` in nanoseconds.

Are you seeing different behavior?

**Comment by SRSaunders** [comment]
Thanks very much for pointing me in the right direction towards Apple's article on `sampleTimestamps:gpuTimestamp:`

The fact that function already takes into account the CPU clock tick interval explains everything.  And yes, MoltenVK performance logging works properly on my M1 machine as well.

Sorry for bothering you with this, but I was looking in the wrong place - at the docs for `MTLTimestamp` which according to Apple is "a timestamp, in Mach absolute time."   Thus I was making an incorrect assumption about the return value units of `sampleTimestamps:gpuTimestamp:`

I appreciate you fixing the issue and spending time to assist my understanding.

**Comment by billhollings** [comment]
> I appreciate you fixing the issue and spending time to assist my understanding.

No worries. You're welcome. Thanks for your investigation and discussion. 

And the questions keep me on my toes! Made me go and confirm.

**Comment by billhollings** [comment]
Closing now.
